### PR TITLE
feat(rules): add PE-005/006, XP-007, REF-003/004 validation rules

### LINKED ISSUE #401: enhancement: new cross-platform and prompt engineering rules (PE-005/006, XP-007, REF-003/004)
## Problem / Motivation

Several cross-platform and prompt engineering validation gaps exist in the current rule set. These rules would catch common anti-patterns that affect agent configuration quality across all supported tools, and detect size/format issues for specific tool requirements.

## Evidence

Searching the codebase confirms none of these rules exist:

- No `PE-005`, `PE-006`, `XP-007`, `REF-003`, or `REF-004` rule IDs in `knowledge-base/rules.json`.
- No validation for redundant/generic instructions, negative-only instructions, AGENTS.md byte limits, duplicate imports, or non-markdown imports.
- The only reference to `REF-003` is in a test comment (`crates/agnix-core/tests/security_integration.rs` line 180) where it appears as a hypothetical - not an implemented rule.

### Prompt Engineering Rules (Universal)

1. **PE-005: Redundant/generic instructions** (WARNING)
   - Detect instructions like "be helpful", "be accurate", "be concise", "follow instructions", "do your best" - these are default LLM behavior and waste context window.
   - Applies to all agent configuration files, not just CLAUDE.md.

2. **PE-006: Negative-only instructions without positive alternative** (WARNING)
   - Detect patterns like "Don't do X" or "Never do Y" without a corresponding "Instead, do Z".
   - Research shows positive framing is more effective for LLM instruction following.
   - Applies universally.

### Cross-Platform Rules

3. **XP-007: AGENTS.md exceeds Codex 32 KiB byte limit** (WARNING)
   - Codex CLI has a documented 32 KiB byte limit for AGENTS.md files.
   - Detect when AGENTS.md file content exceeds 32,768 bytes.
   - Only applies to files named AGENTS.md.

### Reference Rules

4. **REF-003: Duplicate @import** (WARNING)
   - Detect when the same file is imported multiple times in a single file.
   - Duplicate imports waste context window and may indicate copy-paste errors.

5. **REF-004: Import of non-markdown file** (WARNING)
   - Detect `@import` references to non-markdown files (e.g., `.json`, `.py`, `.rs`).
   - While technically valid, importing source code files is usually unintentional and can inject large amounts of non-instructional content.

### Required for each rule

- Add rule definition to `knowledge-base/rules.json` with complete evidence metadata.
- Add human-readable documentation to `knowledge-base/VALIDATION-RULES.md`.
- Add implementation and tests.

## Source

- Codex CLI documentation (XP-007 byte limit)
- Prompt engineering best practices (PE-005, PE-006)
- `crates/agnix-core/src/rules/imports.rs` (REF-003/004 would extend existing import validation)
- `crates/agnix-core/tests/security_integration.rs` line 180 (references REF-003 in a comment)

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary

Implements 5 new validation rules from #401:

- **PE-005**: Redundant generic instructions - detects filler phrases like "be helpful", "be accurate" that waste context window
- **PE-006**: Negative-only instructions - flags "Don't do X" without a positive alternative ("Instead, do Y")
- **XP-007**: AGENTS.md Codex byte limit - warns when AGENTS.md exceeds Codex CLI's 32 KiB limit
- **REF-003**: Duplicate @import - detects the same file imported multiple times (with path normalization)
- **REF-004**: Non-markdown @import - flags imports of .json, .py, .rs files that may not be processed correctly

All rules follow existing patterns: detection in `schemas/`, wiring in `rules/`, i18n in `locales/en.yml`. Total rules: 200 -> 205.

## Changes

- `knowledge-base/rules.json` + `crates/agnix-rules/rules.json` - 5 new rule definitions with evidence metadata
- `crates/agnix-core/src/schemas/prompt.rs` - `find_redundant_instructions()`, `find_negative_only_instructions()`
- `crates/agnix-core/src/rules/prompt.rs` - PE-005/PE-006 validator wiring
- `crates/agnix-core/src/schemas/cross_platform.rs` - `check_byte_limit()`, `CODEX_BYTE_LIMIT`
- `crates/agnix-core/src/rules/cross_platform.rs` - XP-007 validator wiring (AGENTS.md only)
- `crates/agnix-core/src/rules/imports.rs` - REF-003/REF-004 inline checks
- `crates/agnix-core/locales/en.yml` - i18n keys for all 5 rules
- `knowledge-base/VALIDATION-RULES.md` - human-readable documentation
- Rule count updates across README, SPEC, editor docs

## Test plan

- [x] 3,218+ tests pass (`cargo test --workspace`)
- [x] Schema-level unit tests for detection functions (ReDoS protection, code block skipping)
- [x] Validator-level integration tests (config disable, file type scoping)
- [x] Edge cases: empty content, oversized input, path normalization, extensionless imports
- [x] Parity tests pass (rules.json matches VALIDATION-RULES.md)
- [x] CLAUDE.md/AGENTS.md remain identical

Closes #401

### CODE REVIEW

File: crates/agnix-core/src/schemas/prompt.rs [hunk 1]
```diff
@@ -312,6 +314,125 @@ pub fn find_ambiguous_instructions(content: &str) -> Vec<AmbiguousInstruction> {
...
+            continue;
+        }
+
+        if let Some(mat) = pattern.find(line) {
+            results.push(RedundantInstruction {
+                line: line_num + 1,
+                column: mat.start() + 1,
+                phrase: mat.as_str().to_string(),
+            });
+        }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/prompt.rs]: Good catch, applied the suggestion. Now using find_iter() to report all redundant phrases per line. Pushed in 9c06018.

File: crates/agnix-core/src/schemas/prompt.rs [hunk 2]
```diff
@@ -312,6 +314,125 @@ pub fn find_ambiguous_instructions(content: &str) -> Vec<AmbiguousInstruction> {
...
+        if in_code_block {
+            continue;
+        }
+
+        if let Some(mat) = neg_pattern.find(line) {
+            // Check current line and next 2 lines for positive alternative
+            let window_end = (line_num + 3).min(lines.len());
+            let has_positive = lines[line_num..window_end]
+                .iter()
+                .any(|l| pos_pattern.is_match(l));
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/prompt.rs]: Fixed. Window check now skips fenced code blocks. Pushed in cc53ca6.

File: crates/agnix-core/src/rules/imports.rs [hunk 3]
```diff
@@ -104,6 +108,58 @@ impl Validator for ImportsValidator {
...
+                }
+            }
+        }
+
+        // REF-004: Non-markdown @import detection
+        if config.is_rule_enabled("REF-004") {
+            for import in &root_imports {
+                let import_path = Path::new(&import.path);
+                if let Some(ext) = import_path.extension().and_then(|e| e.to_str()) {
+                    if ext != "md" {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/imports.rs]: Fixed. Extension comparison now uses eq_ignore_ascii_case. Pushed in cc53ca6.

File: knowledge-base/rules.json [hunk 4]
```diff
@@ -4376,6 +4480,34 @@
...
+      "severity": "MEDIUM",
+      "category": "cross-platform",
+      "evidence": {
+        "source_type": "vendor_docs",
+        "source_urls": [
+          "https://developers.openai.com/codex/guides/agents-md"
+        ],
+        "verified_on": "2025-05-01",
+        "applies_to": {
+          "tool": "codex-cli"
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/rules.json]: Fixed. Changed to 'codex' to match existing CDX rules. Pushed in 69d155d.

File: website/docs/rules/index.md [hunk 5]
```diff
@@ -205,3 +204,9 @@ This section contains all `200` validation rules generated from `knowledge-base/
 | [AMP-SK-001](./generated/amp-sk-001.md) | Amp Skill Uses Unsupported Field | MEDIUM | Amp Skills | Yes (safe/unsafe) |
 | [RC-SK-001](./generated/rc-sk-001.md) | Roo Code Skill Uses Unsupported Field | MEDIUM | Roo Code Skills | Yes (safe/unsafe) |
 | [XP-SK-001](./generated/xp-sk-001.md) | Skill Uses Client-Specific Features | LOW | Cross-Platform | No |
+| [ROO-001](./generated/roo-001.md) | Empty Roo Code Rule File | HIGH | roo-code | No |
+| [ROO-002](./generated/roo-002.md) | Invalid .roomodes Configuration | HIGH | roo-code | No |
+| [ROO-003](./generated/roo-003.md) | Invalid .rooignore File | MEDIUM | roo-code | No |
+| [ROO-004](./generated/roo-004.md) | Invalid Mode Slug in Rule Directory | MEDIUM | roo-code | No |
+| [ROO-005](./generated/roo-005.md) | Invalid .roo/mcp.json Configuration | HIGH | roo-code | No |
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply website/docs/rules/index.md]: These are auto-generated from rules.json categories. The category field in rules.json uses the slug format consistently. Website display uses a mapping to render them as title case.

File: crates/agnix-core/src/rules/cross_platform.rs [hunk 6]
```diff
@@ -792,4 +815,52 @@ context: fork
...
+
+    #[test]
+    fn test_xp_007_agents_local_md() {
+        let content = "a".repeat(33000);
+        let validator = CrossPlatformValidator;
+        let diagnostics =
+            validator.validate(Path::new("AGENTS.local.md"), &content, &LintConfig::default());
+
+        let xp_007: Vec<_> = diagnostics.iter().filter(|d| d.rule == "XP-007").collect();
+        assert_eq!(xp_007.len(), 1, "XP-007 should apply to AGENTS.local.md");
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/cross_platform.rs]: Fixed. XP-007 now only applies to AGENTS.md (Codex CLI only reads that file). Test updated accordingly. Pushed in 69d155d.

File: crates/agnix-core/src/rules/cross_platform.rs [hunk 7]
```diff
@@ -104,6 +106,26 @@ impl Validator for CrossPlatformValidator {
             }
         }
 
+        // XP-007: AGENTS.md exceeds Codex byte limit (WARNING)
+        if config.is_rule_enabled("XP-007") && is_agents_md {
+            if let Some(exceeded) = check_byte_limit(content, CODEX_BYTE_LIMIT) {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/cross_platform.rs:112]: Fixed. Restricted XP-007 to filename == 'AGENTS.md' only. Pushed in 69d155d.

File: crates/agnix-core/src/schemas/prompt.rs [hunk 8]
```diff
@@ -312,6 +314,125 @@ pub fn find_ambiguous_instructions(content: &str) -> Vec<AmbiguousInstruction> {
...
+        }
+        if in_code_block {
+            continue;
+        }
+
+        for mat in pattern.find_iter(line) {
+            results.push(RedundantInstruction {
+                line: line_num + 1,
+                column: mat.start() + 1,
+                phrase: mat.as_str().to_string(),
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/prompt.rs:361]: Already fixed in 9c06018 - now using find_iter().

File: crates/agnix-core/src/rules/imports.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/imports.rs]: Already fixed in cc53ca6 - now using eq_ignore_ascii_case.

File: crates/agnix-core/src/schemas/prompt.rs [hunk 9]
```diff
@@ -312,6 +314,134 @@ pub fn find_ambiguous_instructions(content: &str) -> Vec<AmbiguousInstruction> {
...
+    }
+
+    results
+}
+
+// ============================================================================
+// PE-006: Negative-Only Instructions
+// ============================================================================
+
+static_regex!(fn negative_only_pattern, r"(?i)^[*\-]?\s*(don't|do not|never|avoid|refrain from)\b");
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/prompt.rs:373]: PE-006 currently targets line-starting negatives which is the most common pattern. Numbered list items (e.g., '1. Don't...') are a good enhancement for a follow-up - tracked.

File: crates/agnix-core/src/rules/imports.rs [hunk 10]
```diff
@@ -104,6 +108,58 @@ impl Validator for ImportsValidator {
...
+        // REF-003: Duplicate @import detection
+        if config.is_rule_enabled("REF-003") {
+            let mut seen_paths: HashSet<String> = HashSet::new();
+            for import in &root_imports {
+                // Normalize: strip leading "./" for comparison
+                let normalized = import
+                    .path
+                    .strip_prefix("./")
+                    .unwrap_or(&import.path)
+                    .to_string();
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/imports.rs:121]: REF-003 currently normalizes ./ prefix which covers the most common duplicate pattern. Deeper canonicalization (e.g., ../dir/../dir) is a good follow-up enhancement.

File: knowledge-base/rules.json [hunk 11]
```diff
@@ -4116,6 +4116,58 @@
       "good_example": "# Project\n\nSee [style guide](docs/STYLE.md) for conventions.\nVisit [GitHub](https://github.com) for source.\n\n(where docs/STYLE.md exists on disk)",
       "bad_example": "# Project\n\nSee [style guide](docs/STYLE.md) for conventions.\nSee [architecture](docs/ARCHITECTURE.md) for design.\n\n(where docs/STYLE.md and docs/ARCHITECTURE.md do not exist)"
     },
+    {
+      "id": "REF-003",
+      "name": "Duplicate Import",
+      "description": "Detects when the same file is @imported multiple times, which wastes context window tokens and may cause confusion.",
+      "severity": "MEDIUM",
+      "category": "references",
+      "evidence": {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/rules.json]: Fixed. Updated References (2->4) and Prompt Eng (4->6) category counts. Pushed in 9d374c2.

File: crates/agnix-rules/rules.json [hunk 12]
```diff
@@ -4116,6 +4116,58 @@
       "good_example": "# Project\n\nSee [style guide](docs/STYLE.md) for conventions.\nVisit [GitHub](https://github.com) for source.\n\n(where docs/STYLE.md exists on disk)",
       "bad_example": "# Project\n\nSee [style guide](docs/STYLE.md) for conventions.\nSee [architecture](docs/ARCHITECTURE.md) for design.\n\n(where docs/STYLE.md and docs/ARCHITECTURE.md do not exist)"
     },
+    {
+      "id": "REF-003",
+      "name": "Duplicate Import",
+      "description": "Detects when the same file is @imported multiple times, which wastes context window tokens and may cause confusion.",
+      "severity": "MEDIUM",
+      "category": "references",
+      "evidence": {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-rules/rules.json]: Same fix as above. Pushed in 9d374c2.

File: website/src/data/siteData.json [hunk 13]
```diff
@@ -1,12 +1,13 @@
...
-  "totalRules": 200,
+  "totalRules": 205,
   "categoryCount": 28,
   "autofixCount": 59,
   "uniqueTools": [
     "amp",
     "claude-code",
     "cline",
     "codex",
+    "codex-cli",
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply website/src/data/siteData.json]: Fixed. Removed codex-cli from uniqueTools. Pushed in 9d374c2.

File: knowledge-base/INDEX.md [hunk 14]
```diff
@@ -1,6 +1,6 @@
 # agnix Knowledge Base - Master Index
 
-> 200 validation rules across 29 categories, sourced from 75+ references
+> 205 validation rules across 29 categories, sourced from 75+ references
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/INDEX.md]: This was a stale count from before merge conflict resolution. Now correctly shows 212 rules across 28 categories. Pushed in 9d374c2.

File: crates/agnix-core/src/schemas/prompt.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/prompt.rs:373]: Acknowledged as enhancement for follow-up. Current implementation covers the core use case.

File: crates/agnix-core/src/schemas/prompt.rs [hunk 15]
```diff
@@ -312,6 +314,134 @@ pub fn find_ambiguous_instructions(content: &str) -> Vec<AmbiguousInstruction> {
...
+static_regex!(fn positive_alternative_pattern, r"(?i)\b(instead|rather|prefer|use\s+\S+\s+instead|better to|should\s+\S+\s+instead)\b");
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/prompt.rs:374]: Acknowledged as enhancement for follow-up. Current implementation covers the core use case.

[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/prompt.rs:374]: Acknowledged as enhancement for follow-up. Current implementation covers the core use case.

[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/prompt.rs:374]: Acknowledged as enhancement for follow-up. Current implementation covers the core use case.

File: test-output.txt [hunk 16]
```diff
@@ -0,0 +1,3409 @@
+    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.23s
+     Running unittests src\main.rs (C:\Users\avife\.cargo\target\debug\deps\agnix-603e48054dc38403.exe)
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply test-output.txt]: Fixed. Removed test-output.txt in 827b613.

File: knowledge-base/VALIDATION-RULES.md [hunk 17]
```diff
@@ -1842,7 +1877,7 @@ pub fn validate_skill(path: &Path, content: &str) -> Vec<Diagnostic> {
 
 ---
 
-****Total Coverage**: 207 validation rules across 28 categories
+****Total Coverage**: 212 validation rules across 28 categories
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/VALIDATION-RULES.md]: The markdown formatting was already corrected in a subsequent commit.
